Adding bind_map. Adding key_error exception.

lyskov opened the pull request:
Adding bind_map which provide minimal bindings for std::map. Adding key_error exception.

wjakob commented:
Hi Sergey,

thank you for the PR. This is in my queue for the next release, but it may take me a bit to review it (I'm leaving for a vacation).

Cheers,
Wenzel

lyskov commented:
Sure thing Wenzel! This is certainly not urgent. And please let me know if there is anything i can do to help with this.

Have a good vacation!

Sergey.

jagerman commented on include/pybind11/stl_bind.h:
```diff
@@ -424,13 +424,18 @@ pybind11::class_<std::map<Key, T, Compare, Allocator>, holder_type> bind_map(pyb
     cl.def("__getitem__",
 		   [](Map &m, const KeyType &k) -> MappedType {
-			   if (m.count(k)) return m[k];
+			   if (m.count(k)) return m.at(k);
```
This could potentially require two lookups--one for the one for the `count`, and one for the element access, and the `at()` is a bit wasteful (as it checks existence again).  The following would avoid both, while being const-usable:

```
auto it = m.find(k);
if (it != m.end()) return *m;
else throw pybind11::key_error();
```

lyskov replied:
Thank you for checking on this! Good point, I will path this!

jagerman replied:
The __delitem__ implementation can do the same thing as well, via `m.erase(it);`

jagerman commented on include/pybind11/stl_bind.h:
```diff
@@ -424,13 +424,18 @@ pybind11::class_<std::map<Key, T, Compare, Allocator>, holder_type> bind_map(pyb
+				   m.erase(r.first);
+				   m.insert( std::make_pair(k, v) );
+			   }
+
```
And using find similarly here cleans up the logic a bit, and avoids needing to erase and reinsert:
```
auto it = m.find(k);
if (it != m.end()) m->second = v;
else m.emplace(k, v);
```

lyskov replied:
Hmm... - this will not work if value type has `const` qualifier, like if map is `map<int, const int>` etc

jagerman replied:
That's basically what I'd *expect* to happen with a const value_type (and is what happens in C++, albeit with a compile time error).

lyskov replied:
Good point! I altered implementation so it now handle const/non-const value differently (otherwise it will be quite confusing for Python user to not be able to alter map's with const values etc)

jagerman commented on include/pybind11/stl_bind.h:
```diff
@@ -364,6 +364,36 @@ NAMESPACE_BEGIN(detail)
+
+}
+
+template<typename Map, typename Class_, typename std::enable_if<std::is_const<typename Map::mapped_type>::value, int>::type = 0>
```
I think you could generalize this a bit by changing
`std::is_const<typename Map::mapped_type>::value`
to
`!std::is_copy_assignable<Map::mapped_type>::value`
which would still catch `const` values, but would also catch non-const types that don't have a copy assignment operator.

lyskov replied:
Good point!

wjakob commented on include/pybind11/stl_bind.h:
```diff
@@ -364,6 +364,36 @@ NAMESPACE_BEGIN(detail)
 template <typename, typename, typename... Args> void map_if_insertion_operator(const Args&...) { }
 
+template <typename Map, typename Class_, typename... Args> void map_if_const_value(Class_ &cl, const Args&...)
+{
```
Indentation issue. Also, this PR suspiciously looks like it's using tabs instead of spaces ;)

lyskov replied:
Fixed! And yes: it was tabs - so i just converted it to spaces.

wjakob commented:
Hi Sergey,

could you rebase this onto master? Also, here are a few more comments:

1. The key iterator has landed, so ``__iter__`` could make use of it.

2. There are various commented parts of the PR -- what about those?

3. It would be nice if there was at least some rudimentary documentation that mentioned the existence of ``stl_bind.h`` in the advanced part of the docs. Right now it's completely undocumented.

THanks,
Wenzel

lyskov commented:
Re make_key_iterator: right, i just pushed the commit that uses it. Please let me know if this is the right way to do it.

@wjakob do i need to register Python testing functions somewhere to make sure that they called during the testing phase? Because right now it does not look like the testing functions is called at all (at least in my setup) (because assert in test_map_string_double should certainly fail due to different numbers of key's, - but it does not). Thought i do see parsing errors if i introduce them into test_stl_binders.py... thoughts?

lyskov commented:
Looks like the tests run fine on appveyor... i guess it is a problem with my setup.

jagerman commented:
> Re make_key_iterator: right, i just pushed the commit that uses it. Please let me know if this is the right way to do it.

The `__iter__` looks good to me (you could write it as: `pybind11::make_key_iterator(m)` if you prefer, but that's purely stylistic).

It would be nice to also add:
```C++
cl.def("items",
    [](Map &m) { return pybind11::make_iterator(m.begin(), m.end()); },
    pybind11::keep_alive<0, 1>() /* Essential: keep list alive while iterator exists */
);
```
to allow pairwise iteration from Python:
```Python
for k, v in mymap.items():
    print("%s = %s" % (k, v))
```

lyskov commented:
Good idea about the 'items' @jagerman ! - I will add this shortly!

lyskov commented:
@wjakob i think the last commit addressed the remaining issues, please let me know if there is anything else that needs to be done. Thanks!

wjakob commented on docs/advanced.rst:
```diff
@@ -550,6 +550,18 @@ and the Python ``list``, ``set`` and ``dict`` data structures are automatically
 out of the box with just the core :file:`pybind11/pybind11.h` header.
 
+Alternatively it might be desirable to bind STL containers as native C++ classes
+therefor eliminating the need of converting back and forth between C++ representation
```
"thereforE" (or even better: just a comma)

wjakob commented on docs/advanced.rst:
```diff
@@ -550,6 +550,18 @@ and the Python ``list``, ``set`` and ``dict`` data structures are automatically
+deal with C++ containers directly instead of using already familiar Python list's
+or dict's.
+
+PyBind11 provide set of binder functions to bind various STL containers like vector's,
```
PyBind11 -> Pybind11
Plural forms use no apostrophe (vector, map)

wjakob commented on docs/advanced.rst:
```diff
@@ -550,6 +550,18 @@ and the Python ``list``, ``set`` and ``dict`` data structures are automatically
+or dict's.
+
+PyBind11 provide set of binder functions to bind various STL containers like vector's,
+map's etc. All binder functions designed to return instances of pybind11::class_
```
... ARE designed to return instances of ``class_``

wjakob commented on docs/advanced.rst:
```diff
@@ -550,6 +550,18 @@ and the Python ``list``, ``set`` and ``dict`` data structures are automatically
+
+PyBind11 provide set of binder functions to bind various STL containers like vector's,
+map's etc. All binder functions designed to return instances of pybind11::class_
+objects so developer could bind extra functions if needed. For complete set of
```
developerS CAN

wjakob commented on docs/advanced.rst:
```diff
@@ -550,6 +550,18 @@ and the Python ``list``, ``set`` and ``dict`` data structures are automatically
+PyBind11 provide set of binder functions to bind various STL containers like vector's,
+map's etc. All binder functions designed to return instances of pybind11::class_
+objects so developer could bind extra functions if needed. For complete set of
+available functions please see :file:`pybind11/stl_bind.h`. For example of usage please
```
For an example on using this feature,

wjakob commented on include/pybind11/stl_bind.h:
```diff
@@ -1,3 +1,4 @@
+// -*- mode:c++;tab-width:4;indent-tabs-mode:f;show-trailing-whitespace:t;rm-trailing-spaces:t -*-
```
remove modeline

wjakob commented on include/pybind11/stl_bind.h:
```diff
@@ -350,4 +354,119 @@ pybind11::class_<std::vector<T, Allocator>, holder_type> bind_vector(pybind11::m
+    cl.def("__setitem__",
+           [](Map &m, const KeyType &k, const MappedType &v) {
+               auto r = m.insert( std::make_pair(k, v) ); // We can't use m[k] = v; because value type might not be default constructable
+               if(!r.second) { // value type might be const so the only way to insert it is to errase it first...
```
space after ``if``

wjakob commented on include/pybind11/stl_bind.h:
```diff
@@ -350,4 +354,119 @@ pybind11::class_<std::vector<T, Allocator>, holder_type> bind_vector(pybind11::m
+    );
+
+	cl.def("items",
+		   [](Map &m) { return pybind11::make_iterator(m.begin(), m.end()); },
```
Spaces, not tabs (please check all files)

wjakob commented on tests/pybind11_tests.cpp:
```diff
@@ -72,6 +73,7 @@ PYBIND11_PLUGIN(pybind11_tests) {
     init_ex_pickling(m);
     init_ex_inheritance(m);
     init_ex_stl_binder_vector(m);
+	init_ex_stl_binder_map(m);
```
spaces, not tabs

lyskov commented:
@wjakob thank you for looking this up! I updated the docs as you suggested and untabifyed the source files.

dean0x7d commented on docs/advanced.rst:
```diff
@@ -979,6 +991,12 @@ exceptions:
 |                                      | accesses in ``__getitem__``, |
 |                                      | ``__setitem__``, etc.)       |
 +--------------------------------------+------------------------------+
+| :class:`pybind11::key_error`         | ``KeyError`` (used to        |
```
Duplicate: `key_error` is already documented a few lines down.

wjakob commented:
This looks good to me now. Two more things: can you remove the duplicate docs pointed out by @dean0x7d and squash this into a single commit?
Thanks, Wenzel

jagerman commented:
Just a heads up: it looks like `stl_bind.h` still has some tabs in it: https://travis-ci.org/pybind/pybind11/jobs/156090804#L287 (that was supposed to start triggering a build failure to let you see it right away, but needs #373 merged to start actually doing that).

lyskov commented:
Ok, this is all should be fixed now.

@wjakob i did squash the commits into one as we'll. (Btw, i think GitHub now have option of doing this automatically when merging PR: after clicking 'merge' on confirmation button there is 'squash and merge' or something)

jagerman commented:
This PR appears to only support `std::map`, but in a lot of cases `std::unordered_map` seems a better choice for C++11 code.  It ought to be very easy to extend this PR to support `unordered_map` by doing something like the following (untested):

```C++
// In pybind11::detail namespace:
template <typename MapType, typename holder_type, typename... Args...> bind_map_type(module &m, const std::string &name, Args&&... args) {
// ... current bind_map() code
}


// In pybind11 namespace:

template <typename Key, typename T, typename Compare = std::less<Key>, typename Allocator = std::allocator<std::pair<const Key, T> >, typename holder_type = std::unique_ptr<std::map<Key, T, Compare, Allocator>>, typename... Args>
pybind11::class_<std::map<Key, T, Compare, Allocator>, holder_type> bind_map(pybind11::module &m, std::string const &name, Args&&... args) {
    return detail::bind_map_type<std::map<Key, T, Compare, Allocator>, holder_type>(m, name, std::forward<Args>(args)...);
}

template <typename Key, typename T, typename Compare = std::less<Key>, typename Allocator = std::allocator<std::pair<const Key, T> >, typename holder_type = std::unique_ptr<std::unordered_map<Key, T, Compare, Allocator>>, typename... Args>
pybind11::class_<std::map<Key, T, Compare, Allocator>, holder_type> bind_unordered_map(pybind11::module &m, std::string const &name, Args&&... args) {
    return detail::bind_map_type<std::unordered_map<Key, T, Compare, Allocator>, holder_type>(m, name, std::forward<Args>(args)...);
}
```

jagerman commented:
That previous post isn't quite right--the unordered_map version needs to take Hash and KeyEqual parameters instead of the Compare parameter, but the idea is the same.

Actually, a nicer alternative might be to just make `bind_map` take the map type, instead of the `Key`/`T`/`Compare`/`Allocator` template parameters, replacing the current bind_map interface with:

```C++
template <typename Map, typename holder_type = std::unique_ptr<Map>, typename... Args>
pybind11::class_<Map, holder_type> bind_map(module &m, const std::string &name, Args&&... args) {
    // ... same body as existing bind_map but without the "using Map = ..." line.
}
```

Then the caller can call it with any compatible map-like class, e.g.
```C++
pybind11::bind_map<std::unordered_map<...>>(m, "Map1");
pybind11::bind_map<std::map<...>>(m, "Map2");
pybind11::bind_map<MyCustomMapType>(m, "Map3");
```

lyskov commented:
@jagerman good idea about generalizing this for different map types! I will keep this in mind for future PR's.

Re passing `map` type vs. passing type arguments: i think i explored that approach initially but in the end decided on passing on individual arguments instead (but not sure now what was the final reason for this). But i do agree that with `map` generalization passing one type might be a more general approach.

wjakob commented:
Hi Sergey,

can you rebase this commit to the master branch? (can't be merged ATM).

Thanks,
Wenzel

wjakob commented:
By the way: if the generalization to unordered maps is really as simple as @jagerman writes above, it would be a pity to not include it in this commit as well.

lyskov commented:
@wjakob i just double checked: - the initial reason for passing arguments for map type is no longer relevant so @jagerman suggestion is now implemented and relevant tests for unordered_map is added. I also refactored bind_vector so it now take Vector type as template parameter instead of pair <T, Allocator> which should allow it to be used with vector-like types. 

This PR is now rebased on current upstream so it is mergeable. Please let me know if there is anything else thats need to be done.

jagerman commented:
The updated vector interface is nice, to keep it consistent with the map interface, but I think that's going to need an extra templated definition for backwards compatibility.

wjakob commented:
This wasn't official API thus far (not documented in any way), so I think it's fine. Furthermore, we are heading towards a major release which can break compatibility with impunity.

(Also, I don't think it's possible to address this incompatibility with an extra template definition in a straightforward way)

lyskov commented:
Great, thank you for taking care of this Wenzel!
